src/SubHask/Algebra/Matrix.hs

```diff
@@ -296,3 +308,206 @@ instance Category (Matrix' vect r) where
     (Mat _) . Zero      = Zero
     (Mat m ) . (Id  r ) = Mat $ m .* r
     (Mat m1) . (Mat m2) = Mat $ mmult m2 m1
+
+
+
+newtype ACCMatrix (bknd::Backend) vect (m::k) (n::k) a = ACCMatrix (A.Acc (A.Array A.DIM2 a))
+
+type ValidACCMatrix b vect m n r =
+  ( FiniteModule vect
+  , r ~ Scalar (Elem vect)
+  , Hilbert vect
+  , VectorSpace r
+  , Prim r
+  , A.Elt r -- A.Elt is the scalar element type; I don't know if this shoud be here but the instance signitures below seem to want it.
```

Comment:
In general, these `ValidXXX` types will have a bunch of other types inside of them.  I'm not too concerned about that.  Just shove whatever you need in here.